addons/l10n_fr_pos_cert/static/src/js/l10n_fr_pos_cert.js

```diff
@@ -0,0 +1,247 @@
+/******************************************************************************
+    Copyright (C) 2017 - Today: GRAP (http://www.grap.coop)
+    @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+    Part of Odoo. See LICENSE file for full copyright and licensing details.
+ *****************************************************************************/
+
+/******************************************************************************
+    Developpers Note: 
+    This module changes the behaviour of the PoS regarding the French Sapin Law.
+    In this context, for french Point of Sales, when a pos order is created
+    a call to the server will be done, to recover the hash of the order to
+    print it on the bill.
+    The time for the bill to print will depend on the server processing time.
+
+    1. If you want to change the message printed on the bill, overload the
+    function prepare_certification_text()
+
+    2. If you want to change the behaviour of the PoS if server is unreachable,
+    (for exemple, prevent printing the bill), overload the following functions :
+    - ReceiptScreenWidget.show_certification()
+    - ProxyDevice.print_receipt_certification()
+******************************************************************************/
+
+'use strict';
+
+odoo.define('l10n_fr_pos_cert.models', function (require) {
+
+    var devices = require('point_of_sale.devices');
+    var models = require('point_of_sale.models');
+    var screens = require('point_of_sale.screens');
+    var DataModel = require('web.DataModel');
+    var core = require('web.core');
+    var _t = core._t;
+
+    /*************************************************************************
+        Promise that will be resolved, when the hash of the saved order is
+        known
+     */
+    var certification_deferred = null;
+
+    /*************************************************************************
+     * Function that return certification text, depending of a hash
+     * @param {string} hash: value of the pos_order.l10n_fr_hash
+     * @param {string} setting: value of the pos_config.l10n_fr_print_hash
+     * @returns {string}: Certification Text that will be printed on the bill
+     */
+    var prepare_certification_text = function(hash, setting){
+        if (setting === 'no_print'){
+            return '';
+        }
+        if (hash){
+            return _t('Certification Number: ') + hash.substring(0, 10) + '...' + hash.substring(hash.length - 10);
+        }
+        return _t("Because of a network problem," +
+            " this ticket could not be certified." +
+            " You can ask later for a certified ticket.");
```

Comment:
is that really true? can we reprint a ticket with the hash later?

Reply:
**In the front office**
- a module exist pos_reprint (in the Odoo core), but AFAIK, it only allows cashier to reprint immediately a receipt, not later. (just the last one)
- I think it's a missing feature. In many shop, the cashier ask "do you want the bill", the customer say no, and the cashier put the bill in the bin. Finally the customer say yes ! 
- For that purpose, I developped a module to do that (to reload done order). See my comment in the main text of the PR : *"The warning message says that the customer can ask later for a certified bill but in the Odoo, it is not possible to print later a done pos.order in Odoo by default. For that purpose, a GRAP V8.0 module is available here https://github.com/grap/grap-odoo-incubator/tree/8.0/pos_done_order_load"*


**in the back office**
We had the possibility to print a pdf (qweb) of a pos order (like a sale order, an invoice, etc...) until V9. 
[Ref](https://github.com/odoo/odoo/blob/9.0/addons/point_of_sale/views/report_receipt.xml). Not available in V10. Do you know why ? If the qweb exist, I can easily add the hash on it.

Reply:
They were removed at 0650047fd59b because they were quite bad. However, I do agree we need to be able to reprint past tickets from the pos interface. we were talking not long ago with @lap-odoo and hopefully, in v13 we can get this.

Reply:
V13 ! OK.
- So, immediately, I propose to change the text for a more generic one. *"Because of a network problem, this ticket could not be certified."*
- When you'll talk again with @lap-odoo, about the reprint feature, please consider that some shops now propose to send the bill by email too. (The qweb report can be interesting for that).